ensime_shared/spec/features/terrain.py

```diff
@@ -0,0 +1,115 @@
+"""
+Shared test setup such as setup/teardown hooks and fixtures.
+
+This file is loaded automatically by Lettuce, see:
+    http://lettuce.it/reference/terrain.html
+"""
+
+import os
+import shutil
+import sys
+import tempfile
+from contextlib import contextmanager
+from os import path
+
+import neovim
+from lettuce import after, before, world
+
+DEBUG = os.environ.get('DEBUG')
+LISTEN_ADDRESS = os.environ.get('NVIM_LISTEN_ADDRESS')
+
+FEATURESROOT = path.dirname(path.abspath(__file__))
+
+neovim.setup_logging()
+
+
+# TODO: Will we need a @before.each_feature that nukes .ensime_cache, etc.?
+@before.all
+def setup_test_environment():
+    world.project_path = path.join(tempfile.mkdtemp(), 'testproject')
+    world.bootstrap_projects_path = tempfile.mkdtemp()
+
+    resources = path.realpath(path.join(FEATURESROOT, '..', 'resources'))
+    shutil.copytree(path.join(resources, 'testproject'), world.project_path)
+    world.vimrc = path.join(resources, 'vimrc')
+
+
+@after.all
+def teardown_test_environment(_total):
+    if DEBUG:
+        print('-' * 80)
+        print('Project: ' + world.project_path)
+        print('Bootstraps: ' + world.bootstrap_projects_path)
+        print('Left for inspection.')
+        print('-' * 80)
+    else:
+        world.vim.quit()  # Implicitly shut server down
+        shutil.rmtree(world.project_path)
+        shutil.rmtree(world.bootstrap_projects_path)
+
+
+@before.each_scenario
+def reset_vim(_scenario):
+    # Boilerplate to set up embedded Neovim test instance, by default, allowing
+    # separate process optionally.
+    if LISTEN_ADDRESS:
+        world.vim = neovim.attach('socket', path=LISTEN_ADDRESS)
+    else:
+        embed_cmd = ['nvim', '-u', world.vimrc, '-i', 'NONE', '--embed']
+        world.vim = neovim.attach('child', argv=embed_cmd)
+
+    world.vim.input(CLEANUP_FUNC.format(vimrc=world.vimrc))
```

Comment:
This is mostly where rough edges surface.

The [neovim/python-client test suite][2] which I pillaged for ideas sets up the `vim` instance *once* at the beginning of the suite, and calls the state reset function per-test. I tried this, and it worked fine in non-embedded mode, but with embedding it just hung consistently, very early in the test run. I didn't figure out why yet. Changing it to `@before.each_scenario` worked around that. Lettuce's `world` is shared global state, so it might be problematic for concurrent code. It would of course be more performant to do it once, though with only a few tests so far I believe it's negligible.

The Python client's tests also don't need to test any plugins like ours do, so there are some differences. AFAIK there is currently no way to *reload* a remote plugin in a Neovim instance, like you might be able to do with a VimL plugin by un-`let`ing include guard vars, removing autocommands, and doing `runtime plugin/foo.vim` afterward. I asked in Neovim Gitter about this. Making a new `vim` instance per-scenario as done here probably obviates the need for that anyway, but I'm not entirely confident of my understanding of the whole remote plugin host architecture yet, so I'm not certain.

I also have a hunch that changing out the `vim` instance like this might be a source of the ENSIME server processes getting left behind after test runs—perhaps Python garbage collection steps in without the `VimLeave` getting run.

I've left the `BeforeEachTest` function infrastructure in place here in case we can figure out a way to move the `vim` instance setup back into `@before.all`, maybe others want to take a stab at it when trying this branch. It is probably unnecessary as it stands now, with the new `vim` instance per-test.

Also worth mentioning, there is [a `remote#host#LoadRemotePlugins` function][1] in the Neovim API. When I had the `vim` instantiation in `@before.all`, I tried calling that in the `BeforeEachTest` function at the end after re-`sourcing` the vimrc, and that worked to accomplish sanitizing the session state and then loading our plugin, but *only when running a single scenario* with `lettuce -s [N]`. Other code in that Neovim API throws an error when a remote plugin gets "registered" more than once when executing multiple tests 😞  This is an area where the Neovim team is probably still open to ideas, but of course it's not likely to change overnight.

[1]: https://github.com/neovim/neovim/blob/7fa1baf44e7877b7b6b861c7140b006d4cee79f3/runtime/autoload/remote/host.vim#L130
[2]: https://github.com/neovim/python-client/blob/master/test/test_common.py